### PR TITLE
PathContainer#clearAndShowPath, History#peek(int)

### PR DESCRIPTION
**Pull request by rjrjr**

### CODE REVIEW

File: flow-path/src/main/java/flow/path/PathContainer.java [hunk 1]
```diff
@@ -68,9 +68,29 @@ protected PathContainer(int tagKey) {
...
+    doShowPath(view, traversal.destination.<Path>top(), traversal.direction, viewState, callback);
+  }
+
+  /**
+   * Replace the current view and show the given path. Allows display of {@link Path}s other
+   * than in response to Flow dispatches.
+   */
+  public void clearAndShowPath(PathContainerView view, Path path, Flow.Direction direction,
+      Flow.TraversalCallback callback) {
+    ViewState nullViewState = new ViewState() {
```
Comment (loganj) [inline flow-path/src/main/java/flow/path/PathContainer.java]: constant

Comment (rjrjr) [inline flow-path/src/main/java/flow/path/PathContainer.java]: Fixed.

File: flow/src/main/java/flow/History.java [hunk 2]
```diff
@@ -121,20 +120,26 @@ public int size() {
 
   public <T> T top() {
     //noinspection unchecked
-    return (T) history.peek().state;
+    return (T) peek(0);
+  }
+
+  public <T> T peek(int index) {
```
Comment (rjrjr) [inline flow/src/main/java/flow/History.java]: Should we add the same method to builder?

Comment (loganj) [inline flow/src/main/java/flow/History.java]: Couldn't hurt.

File: flow-path/src/main/java/flow/path/PathContainer.java [hunk 3]
```diff
@@ -27,6 +27,8 @@
  */
 public abstract class PathContainer {
 
+  public static final ViewState NULL_VIEW_STATE = new NullViewState();
```
Comment (loganj) [inline flow-path/src/main/java/flow/path/PathContainer.java]: public?

File: flow-path/src/main/java/flow/path/PathContainer.java [hunk 4]
```diff
@@ -68,9 +70,22 @@ protected PathContainer(int tagKey) {
...
-    Path path = traversal.destination.top();
     ViewState viewState = traversal.destination.currentViewState();
+    doShowPath(view, traversal.destination.<Path>top(), traversal.direction, viewState, callback);
+  }
+
+  /**
+   * Replace the current view and show the given path. Allows display of {@link Path}s other
+   * than in response to Flow dispatches.
+   */
+  public void clearAndShowPath(PathContainerView view, Path path, Flow.Direction direction,
```
Comment (loganj) [inline flow-path/src/main/java/flow/path/PathContainer.java]: Maybe just `setPath`?

File: flow-path/src/main/java/flow/path/PathContainer.java [hunk 5]
```diff
@@ -100,4 +115,12 @@ private TraversalState ensureTag(ViewGroup container) {
     }
     return traversalState;
   }
+
+  private static class NullViewState implements ViewState {
```
Comment (loganj) [inline flow-path/src/main/java/flow/path/PathContainer.java]: final

**Comment by loganj** [comment]
A few notes, LG otherwise.

File: flow/src/main/java/flow/History.java (hunk 2 above)
Comment (rjrjr) [inline flow/src/main/java/flow/History.java]: Meh, then History has top() / peek(int), and builder would have peek() / peek(int). Seems like we need to make up our minds. I'll leave Builder alone for now.

File: flow-path/src/main/java/flow/path/PathContainer.java (hunk 3 above)
Comment (rjrjr) [inline flow-path/src/main/java/flow/path/PathContainer.java]: Oops.

File: flow-path/src/main/java/flow/path/PathContainer.java (hunk 4 above)
Comment (rjrjr) [inline flow-path/src/main/java/flow/path/PathContainer.java]: :+1:

**Pull request merged**